Close #68 immutable map across code

raulccabreu opened the pull request:
and some other improvements (custom adapter, new state rendering and tests).

- Change the use of lists for Gifs (State) to Map and perform necessary changes across code.
- Change gifs type on AppState: from SortedMap/TreeMap to Map/LinkedHashMap. Sorted maps are automatically ordered by keys, so linked hash maps (keep insertion order) are the wanted type.
- Change loadGifs location from onCreate to onStart and now it is called after stateListener setting.
- Change hasStateChanged implementations to use classes equals to simply compare states change or other some specifically state comparison.
- Create GifsAdapter following SimpleAdapter implementation from Anvil Recycler but with some custom changes to improve our usage.
- Minimize Anvil.render calls by removing/change forced onStateChange moments, making stateChanged verification to be more specifically and calling Anvil.render on some specifically scenarios. (unnecessary render calls are creating undesirables bugs after new immutable and custom adapter changes)

Changes needed on tests after immutable, adapter and fixes:

- Change `withEqualsGifDrawable` to `withEqualsGifUuid` to using on `whenListKeepOrderedAfterPlayTest`
- Change `actionOnItemAtPosition(lastAdapterPos, click())` to `withRecyclerPos(lastScreenPos)).perform(click()` because we don't need the scrolling of first option (the scrolling was made before to load gifs on time)
- Remove the scrolling to top code on `whenScrollUpAndPlayFirstItemTest` because `actionOnItemAtPosition(0, click())` scrolls to top and click.
-  Fixing and improving `whenListKeepOrderedAfterPlayTest`

raulccabreu committed:
Close #68 (immutable map across code) and some other improvements (custom adapter, new state rendering and tests)

- Change the use of lists for Gifs (State) to Map and perform necessary changes across code.
- Change gifs type on AppState: from SortedMap/TreeMap to Map/LinkedHashMap. Sorted maps are automatically ordered by keys, so linked hash maps (keep insertion order) are the wanted type.
- Change loadGifs location from onCreate to onStart and now it is called after stateListener setting.
- Change hasStateChanged implementations to use classes equals to simply compare states change or other some specifically state comparison.
- Create GifsAdapter following SimpleAdapter implementation from Anvil Recycler but with some custom changes to improve our usage.
- Minimize Anvil.render calls by removing/change forced onStateChange moments, making stateChanged verification to be more specifically and calling Anvil.render on some specifically scenarios. (unnecessary render calls are creating undesirables bugs after new immutable and custom adapter changes)

Changes needed on tests after immutable, adapter and fixes:

- Change `withEqualsGifDrawable` to `withEqualsGifUuid` to using on `whenListKeepOrderedAfterPlayTest`
- Change `actionOnItemAtPosition(lastAdapterPos, click())` to `withRecyclerPos(lastScreenPos)).perform(click()` because we don't need the scrolling of first option (the scrolling was made before to load gifs on time)
- Remove the scrolling to top code on `whenScrollUpAndPlayFirstItemTest` because `actionOnItemAtPosition(0, click())` scrolls to top and click.
-  Fixing and improving `whenListKeepOrderedAfterPlayTest`

raulccabreu committed:
Close #67: using onWindowsFocusChanged to register/unregister StateListener

- Inspired by (http://stackoverflow.com/a/38966832) and after some tests onWindowsFocusChanged was the best fit solution for the new local of registerOnStateChange and unregisterOnStateChange. This new solution makes UI receives the stateChange only when necessary.

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/GifListActivity.java:
```diff
@@ -39,11 +38,25 @@ protected void onCreate(Bundle savedInstanceState) {
-        if (!appState.getGifs().isEmpty()) {
+        if (!appState.getGifs().isEmpty() && mGifProgressVisibility) {
             mGifProgressVisibility = false;
+            anvilRender();
```
Why only render when progress is visible? If changes happens on screen ?

raulccabreu replied:
In the activity actual scenario i put render inside to minimize unnecessary render because other components listen to StateChange and render when needed too. But move render to outside any condition on StateChange is a option to make easy our future code maintenance where we will not need any adjustment on render local. What do you think?

niltonvasques replied:
Yes that is the idea! In futures maintenance it`s better to have the things working without too much effort!

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/components/FeedComponent.java:
```diff
@@ -58,29 +57,33 @@ public FeedComponent(Context context, AttributeSet attrs, int defStyleAttr) {
+                mGifsAdapter = gifsAdapter(mGifs, (gif) -> renderGifView(gif));
+                Recycler.adapter(mGifsAdapter);
+            }
+            else if (!mGifs.equals(mGifsAdapter.getGifs())) {
```
This else if pattern looks weird to me.
I prefer the common and straightforward way:

```java
if (....) {
   //...
} else if (.....) {
   //...
}
```

But it's ok! However we need to define a pattern to be used over all code, since I believe that there is different else if style along the code.

raulccabreu replied:
You are correct and the pattern is this you saying. I don't know why code inspect don't catch this.

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/components/GifComponent.java:
```diff
@@ -102,29 +105,27 @@ public void setAnvilRenderListener(AnvilRenderListener listener) {
+        else {
+            unregisterOnStateChange();
+        }
+    }
```
Duplicated code. Maybe we can extends RenderableView class to add this behavior. What do you think?

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/components/GifsAdapter.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public void setGifs(Map<String, Gif> gifs) {
+        mGifs = ImmutableMap.copyOf(gifs);
```
Why make a copy?

raulccabreu replied:
Firstly, I tried to resolve that problem of two activities using the same state on adapter, but now thinking with more attention, it isn't the solution... :(, the value can be changed on exact time of using inside view...the unique solution maybe a synchronized one. You have this opinion too?

niltonvasques replied:
Now with only one adapter living not solves the problem?

raulccabreu replied:
Maybe...

fpbitencourt commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/components/FeedComponent.java:
```diff
@@ -58,29 +57,33 @@ public FeedComponent(Context context, AttributeSet attrs, int defStyleAttr) {
+            }
+            else if (!mGifs.equals(mGifsAdapter.getGifs())) {
+                mGifsAdapter.setGifs(mGifs);
+                mGifsAdapter.notifyDataSetChanged();
```
Chamar _notifyDataSetChanged_ para qualquer alteração da lista de gif pode ser problemático. O ideal seria chamar somente caso haja alterações em diversos itens da lista, para alterações menores, em itens específicos, deveríamos utilizar _notifyItemChanged_. Possivelmente teremos problemas de desempenho de uso geral da lista, principalmente quando tiver animações.

fpbitencourt commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/components/FeedComponent.java:
```diff
@@ -58,29 +57,33 @@ public FeedComponent(Context context, AttributeSet attrs, int defStyleAttr) {
                 mLayoutManager = new LinearLayoutManager(getContext(),
                         LinearLayoutManager.VERTICAL, false);
             }
             Recycler.layoutManager(mLayoutManager);
```
Uma dúvida, o trecho de código abaixo poderia ficar dentro do _if_ que é instanciado o _mLayoutManager_ ou precisam ser chamados sempre que a view é renderizada?

```
Recycler.layoutManager(mLayoutManager);
Recycler.hasFixedSize(true);
```

niltonvasques replied:
Acredito que deve ser chamado sempre que a view é renderizada virtualmente. Por que uma chamada ao método view não necessariamente vai redesenhar a view na tela. Caso ela esteja dentro do if, nas próximas vezes que passar no método view e a Recycler for recriada virtualmente, ela não vai conter um layout manager nem o fixedSize true.

raulccabreu committed:
- Add checkstyle check to verify right brace on the same line of next word and fixes the errors informed after this change.
- Create the class RenderableComponent with base code between GifComponent and FeedComponent.
- Change the adapter gifs copy to a simple set.
- Fix the adapter getItemViewType implementation.
- Move the layoutManager, adapter and hasFixedSize anvil setters to outside the If's to always be called when anvil render.
- Move the Anvil.render calls from OnStateChange of Activity, Gif and Feed Components to be called after all OnStateChange calls.
- Implement on GifComponenet a logic to only calls ActionCreator.gifDownloadStart only time per Gif until fails or GifComponent changes the binded gif.

raulccabreu commented:
Changes made, i fix one major problem that was related with multiple gif download start or any one shot action that we will make inside anvil view methods. With Anvil.render inside each specific stateListener view (gif, feed, activity) the first one start the render of others and so on...even when the others components state was not changed. My fix move the Anvil.Render to be made after all onStateChanged calls, and this render components with a less hunger approach. With other specific Anvil.Render calls (not stateChanged ones) we need to be cautious about desirables one shot actions inside view method.

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/MyDispatcherImpl.java:
```diff
@@ -0,0 +1,27 @@
+    @Override
+    protected void notifyStateListeners(AppState newState, AppState oldState) {
+        super.notifyStateListeners(newState, oldState);
+        Anvil.render();
```
Looks a nice solution :D

niltonvasques commented on app/src/main/java/br/com/catbag/gifreduxsample/ui/GifListActivity.java:
```diff
@@ -74,6 +65,8 @@ private void bindingViews() {
                 visibility(mGifProgressVisibility);
             });
+
+            if (mAnvilRenderListener != null) mAnvilRenderListener.onAnvilRendered();
```
Now that we have a custom Dispatcher that calls `Anvil#render()`, looks easy to move anvilRenderedListener to there. What do do you think? But I don't know if Anvil.render() is a sync or async method.

raulccabreu replied:
Anvil.Render is sync on ui thread. But this will not catch others rendering cases (with no state change at all), like onClick button, or adapters scrolling rendering, etc. Or some specific Anvil.Render call that we have to make.

niltonvasques commented:
:+1:
